ships/0028-buildrun-cleanup.md

```diff
@@ -0,0 +1,152 @@
+
+A successful BuildRun is a BuildRun where the `type=Succeeded` condition is set to `"true"`. A failed BuildRun has the `type=Succeeded` condition is set to `"false"`, and contains therefore for example canceled BuildRuns as well.
+
+The two TTL fields behaves the same as the [`ttlSecondsAfterFinished` field for Jobs](https://kubernetes.io/docs/concepts/workloads/controllers/ttlafterfinished/). I decided against naming consistency with that feature, in favor of consistency inside our resource. Builds and BuildRuns already have a `timeout` field where we also use a duration while Kubernetes Jobs use an integer to store just seconds. Durations are easier to specify for users because they can simply write something like `6h` and don't need to calculate the amount of seconds.
+
+The `successfulLimit` and `failedLimit` fields hold a positive integer number. They define how many successful and failed BuildRuns for a Build can exist.
+
+If both TTL and limit are defined, then the BuildRun will get deleted once the first criteria is met.
+
+The BuildRun resource will be extended with a `retention` section with only the `ttlAfterFailed`, and `ttlAfterSuccessful` fields. They override the values of the Build. This is done for consistency (the `timeout` can also be overwritten in the BuildRun) and in support of standalone BuildRuns that we may have in the future.
+
+### Backend
+
+When a BuildRun completes (condition `type=Succeeded` is changed from `"unknown"` to `"true"` or `"false"`), we will run additional logic:
+
+We take `successfulLimit`, or `failedLimit` from the BuildRun's `.status.buildSpec` (TBD: is this good enough or does the user assume a potentially updated value will be taken from the Build ? This is related to the TBD at the end of this section). If any of them is defined, then ...
+
+- We will run a list for BuildRuns in the namespace with the label filter `build.shipwright.io/name: <BUILD_NAME>`.
+- All BuildRuns will be put into three buckets: successful (condition `type=Succeeded` set to `"true"`), failed (condition `type=Succeeded` set to `"false"`) and the rest.
+- If the number of successful BuildRuns is larger than `successfulLimit`, then the oldest successful BuildRun(s) sorted by `.metadata.creationTimestamp` will be deleted.
+- If the number of failed BuildRuns is larger than `failedLimit`, then the oldest failed BuildRun(s) sorted by `.metadata.creationTimestamp` will be deleted.
+
+We can consider to run above logic in a go routine rather than as part of the BuildRun reconciliation. Failures would be ignored which is okay.
+
+When the `ttlAfterFailed` field is set for a BuildRun that failed, or when the `ttlAfterSuccessful` field is set for a BuildRun that succeeded, then we will return a reconciliation result where `Requeue` is set to `true` and `RequeueAfter` is set to `.status.completionTime + TTL - now`.
```

Comment:
any reasons why you didnt choose the `lasttransitiontime` of our `Succeeded` condition. I believe is more accurate as it based on our controller updates(last state transition)

Reply:
Makes sense, had forgotten that we take `completionTime` from the TaskRun.

Reply:
Done.